### ISSUE TITLE
URL slash encoded

### ISSUE THREAD

**Issue by jules-ch** [tiangolo/fastapi#791]
### Describe the bug

FastAPI is not routing URL containing slash encoded.

### To Reproduce

1. Create a file with:

```Python
from fastapi import FastAPI
app = FastAPI()


@app.get("/units/{unit}")
def read_unit(unit: str):
    return {"unit": unit}

```

3. Open the browser and call the endpoint `/units/kg%2Fs`.
4. It returns Status 404 with JSON with `{"detail":"Not Found"}`.
5. But I expected it to return `{"unit": "kg/s"}`.
6. Working with `/units/kg`, returning `{"unit": "kg"}`.

### Expected behavior

FastAPI working with URL encoded slash

### Environment

- OS:  Windows
- FastAPI Version : 0.45.0
- Python 3.7.3

### Additional context

Starlette regexp seems fine though when testing it returns this route regexp: 

``` regexp
'^/units/(?P<unit>[^/]+)$'
```
```python
>>> import re
>>> pattern=re.compile('^/units/(?P<unit>[^/]+)$')
>>> pattern.match('/units/kg%2Fs')
<re.Match object; span=(0, 13), match='/units/kg%2Fs'>
>>> pattern.match('/units/kg%2Fs').groupdict()["unit"]
'kg%2Fs'
```

**Comment by dmontagu** [comment]
Seems like this is probably an issue with starlette -- I'd maybe try to create a reproducible example using only starlette (should be very similar to what you've done already) and create an issue there.

For what it's worth, it would also be easier to test using the starlette `TestClient` -- that way you wouldn't need to access the endpoint in a browser.

**Comment by jules-ch** [comment]
Yeah, tried it with the latest version of starlette & I encounter the same issue. I'll put an issue in their tracker.

**Comment by tiangolo** [comment]
Thanks @jules-ch and @dmontagu !

Then let's track that in Starlette, right?

**Comment by jules-ch** [comment]
@tiangolo Sorry it was on my todo at some point & I completely forgot.
The issue has been transfered to the starlette repo tracker.

**Comment by jules-ch** [comment]
As stated there https://github.com/encode/starlette/issues/826#issuecomment-593124738.

It seems to be starlette related linked with the current ASGI spec.
The path is decoded & the routing is made on this decoded path.

The ASGI spec has implemented a `raw_path` parameter but for now its better to stick to query parameters for those kind of use cases.

I'm closing it for now.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue @jules-ch ! :+1:

**Comment by fsonntag** [comment]
I'm running in the same issue. Using a query parameter is not really feasible in all cases or a proper way to build an API though. I hope some changes from the ASGI spec updates can be streamlined into Starlette/FastAPI.

**Comment by Gatsby-Lee** [comment]
@jules-ch  
so, how should I deal with this?

**Comment by jules-ch** [comment]
@Gatsby-Lee 

If you want to keep this kind of behavior with slash in your routing path.
You can use this workaround with a specific character to split your url, like `-` or `#`.
We can do that with starlette with the routing convertors `path` (https://www.starlette.io/routing/).

Keeping my example, you can try the following:

```python

from fastapi import FastAPI

app = FastAPI()


@app.get("/units/-/{unit:path}")
def read_unit(unit: str):
    return {"unit": unit}
```

You can now visit the URL `http://127.0.0.1:8000/units/-/kg/s` which will outputs : 

```json
{"unit":"kg/s"}
```

**Comment by jules-ch** [comment]
I know Gitlab is using this to route their subgroups & repositories.

**Comment by Gatsby-Lee** [comment]
@jules-ch  I see. Thank you for your comment. I will try what you shared :)

**Comment by Menotomy** [comment]
@jules-ch
I came across this issue and tried your suggestion for something similar and it worked. My use case was returning the name of a specific interface in a network router. Network routers have names with slashes, like et-1/0/0 (Ethernet HundredGig in this case). This means the URL needs to look like et-1%2f0%2f0, as %2f is for a / character. 

With the route
```python
@app.get('/cfg/interfaces/interface/{int_name}')
def get_int(int_name: str):
```
the URL sent to FastAPI would be http://1.2.3.4:8000/cfg/interfaces/interface/et-1/0/0 which is an incorrect route resulting in a 404 Not Found. 

I followed your suggestion with adding `:path` and the route worked.
```python
@app.get('/cfg/interfaces/interface/{int_name:path}')
def get_int(int_name: str):
```
The function was able to parse XML using xpath, using that variable for the interface name. 

Just confirming and providing an additional example of anyone that sees this.

**Comment by jules-ch** [comment]
Good to hear, be careful to split your URL with a characters that are not encoded in the URI see https://tools.ietf.org/html/rfc3986#section-2.2 & https://tools.ietf.org/html/rfc3986#section-2.3 to make your url path unique especially if you want to add more URLs after `/cfg/interfaces/interface`.
If not you can keep it that way.

**Comment by basiltt** [comment]
I wanted to send string which contains backslash, I have managed to do it this way. After accepting the data I have converted it as mentioned below.

```
class LazyDecoder(json.JSONDecoder):
    def decode(self, s, **kwargs):
        regex_replacements = [
            (re.compile(r'([^\\])\\([^\\])'), r'\1\\\\\2'),
            (re.compile(r',(\s*])'), r'\1'),
        ]
        for regex, replacement in regex_replacements:
            s = regex.sub(replacement, s)
        return super().decode(s, **kwargs)


@router.post("/add_secret", status_code=status.HTTP_201_CREATED)
async def add_secret(background_tasks: BackgroundTasks, secret_raw: SecretCreationForm,
                     session: AsyncSession = Depends(get_db)):
    secret = json.loads(json.dumps(secret_raw.dict()), cls=LazyDecoder)
```

**Comment by swamidass** [comment]
So I am having this issue arise in a use case where I cannot switch to query parameters. What is the right fix?

**Comment by iudeen** [comment]
> So I am having this issue arise in a use case where I cannot switch to query parameters. What is the right fix?

From comments above:
```
@app.get('/cfg/interfaces/interface/{int_name:path}')
def get_int(int_name: str):
```
